### PR TITLE
React video sharing dialog

### PR DESCRIPTION
**Pull request by Calinteodor**

### CODE REVIEW

**Comment by saghul** [comment]
Jenkins add to whitelist.

**Comment by saghul** [comment]
Jenkins please test this please.

File: react/features/shared-video/actions.js [hunk 1]
```diff
@@ -1,16 +1,19 @@
-import { SET_SHARED_VIDEO_STATUS, TOGGLE_SHARED_VIDEO } from './actionTypes';
+// @flow
```
Comment (saghul) [inline react/features/shared-video/actions.js:1]: Let's rename this to actions.web.js while we merge them.

File: react/features/shared-video/actions.js [hunk 2]
```diff
@@ -1,16 +1,19 @@
...
+
+import { openDialog } from '../base/dialog';
```
Comment (saghul) [inline react/features/shared-video/actions.js:3]: Please import from .../base/dialog/actions

File: react/features/shared-video/components/_.web.js [hunk 3]
```diff
@@ -0,0 +1,3 @@
+// @flow
```
Comment (saghul) [inline react/features/shared-video/components/_.web.js:1]: you can drop the flow comment here

File: react/features/shared-video/components/index.js [hunk 4]
```diff
@@ -0,0 +1,3 @@
+// @flow
```
Comment (saghul) [inline react/features/shared-video/components/index.js:1]: ditto

File: react/features/shared-video/components/native/SharedVideoButton.js [hunk 5]
```diff
@@ -0,0 +1,117 @@
+// @flow
+
+import type { Dispatch } from 'redux';
+
+import { isSharingStatus } from '../../';
```
Comment (saghul) [inline react/features/shared-video/components/native/SharedVideoButton.js:5]: specify the file

File: react/features/shared-video/components/native/SharedVideoButton.js [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+
+    if (ownerId !== localParticipantId) {
+        return {
+            _isDisabled: isSharingStatus(sharedVideoStatus),
+            _sharingVideo: false,
+            visible };
+    }
+
+    return {
+        _sharingVideo: isSharingStatus(sharedVideoStatus),
```
Comment (saghul) [inline react/features/shared-video/components/native/SharedVideoButton.js:112]: let's set _isDisabled to false here

File: react/features/shared-video/components/web/SharedVideoDialog.js [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+                width = { 'small' }>
+                <FieldTextStateless
+                    autoFocus = { true }
+                    compact = { false }
+                    label = { t('dialog.videoLink') }
+                    onChange = { this._onChange }
+                    placeholder = { defaultSharedVideoLink }
+                    shouldFitContainer = { true }
+                    type = 'text'
+                    value = { value } />
```
Comment (saghul) [inline react/features/shared-video/components/web/SharedVideoDialog.js:44]: shouldn't this be `this.state.inputValue` ?

File: react/features/shared-video/components/AbstractSharedVideoDialog.js [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+    /**
+     * Validates the entered video link by extracting the id and dispatches it.
+     *
+     * It returns a boolean to comply the Dialog behaviour:
+     *     {@code true} - the dialog should be closed.
+     *     {@code false} - the dialog should be left open.
+     *
+     * @returns {boolean}
+     */
+    _onSetVideoLink() {
```
Comment (saghul) [inline react/features/shared-video/components/AbstractSharedVideoDialog.js:103]: On native this function will take a string, which is the value of the field.

File: react/features/shared-video/components/AbstractSharedVideoDialog.js [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+
+    _onChange: Object => void;
+
+    /**
+     * Callback for the onChange event of the field.
+     *
+     * @param {Object} evt - The static event.
+     * @returns {void}
+     */
+    _onChange(evt) {
```
Comment (saghul) [inline react/features/shared-video/components/AbstractSharedVideoDialog.js:83]: This applies to web only.

File: react/features/shared-video/components/web/SharedVideoDialog.js [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+    render() {
+        const { t, value } = this.props,
+            { okDisabled } = this.state;
+
+        return (
+            <Dialog
+                hideCancelButton = { false }
+                okDisabled = { okDisabled }
+                okKey = { t('dialog.Share') }
+                onSubmit = { this._onSetVideoLink }
```
Comment (saghul) [inline react/features/shared-video/components/web/SharedVideoDialog.js:33]: use a local function, extract the input value and then call onSetVideoLink

File: react/features/shared-video/components/native/SharedVideoDialog.js [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+     *
+     * @inheritdoc
+     */
+    render() {
+        return (
+            <InputDialog
+                contentKey = 'dialog.shareVideoTitle'
+                onSubmit = { this._onSetVideoLink }
+                textInputProps = {{
+                    placeholder: 'https://youtu.be/TB7LlM4erx8'
```
Comment (saghul) [inline react/features/shared-video/components/native/SharedVideoDialog.js:24]: Use the constant

File: react/features/shared-video/reducer.js [hunk 12]
```diff
@@ -1,3 +1,5 @@
+// @flow
```
Comment (saghul) [inline react/features/shared-video/reducer.js:1]: Rename this to reducer.web

File: react/features/toolbox/components/web/Toolbox.js [hunk 13]
```diff
@@ -71,6 +70,7 @@ import {
     OverflowMenuVideoQualityItem,
     VideoQualityDialog
 } from '../../../video-quality';
+import { toggleSharedVideo } from '../../../youtube-player/actions';
```
Comment (saghul) [inline react/features/toolbox/components/web/Toolbox.js:73]: Please move the action to shared-video. Let's get rid off youtube-player altogether

File: react/features/toolbox/components/web/Toolbox.js [hunk 14]
```diff
@@ -1056,12 +1056,10 @@ class Toolbox extends Component<Props, State> {
                     key = 'record'
                     showLabel = { true } />,
             this._shouldShowButton('sharedvideo')
-                && <OverflowMenuItem
+                && <SharedVideoButton
                     accessibilityLabel = { t('toolbar.accessibilityLabel.sharedvideo') }
```
Comment (saghul) [inline react/features/toolbox/components/web/Toolbox.js:1060]: I think the label is set in the component.

File: react/features/youtube-player/functions.js [hunk 15]
```diff
@@ -5,7 +5,7 @@ import { getParticipants } from '../base/participants';
 import { YOUTUBE_PARTICIPANT_NAME } from './constants';
 
 /**
- * Returns true if there is a youtube video being shaerd in the meeting.
+ * Returns true if there is a youtube video being shared in the meeting.
```
Comment (saghul) [inline react/features/youtube-player/functions.js:8]: Let's remove youtube-player altogether and merge it with shared video.

**Review by saghul — CHANGES_REQUESTED** [review]

File: react/features/shared-video/components/web/SharedVideoDialog.js [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+                titleKey = { t('dialog.shareVideoTitle') }
+                width = { 'small' }>
+                <FieldTextStateless
```
Comment (saghul) [inline react/features/shared-video/components/web/SharedVideoDialog.js:36]: Set the `name` prop to the previous name so the torture tests pass.

Comment (Calinteodor) [reply react/features/shared-video/components/web/SharedVideoDialog.js:82]: All the review remarks have been implemented @saghul. Thank you!

File: package-lock.json [hunk 17]
```diff
@@ -1050,17 +1050,17 @@
       "integrity": "sha512-x5HFsW+E/nQalGMw7hu+fvPqnBeBaIr0lWJ2SG0PPL2j+Pm9lYvCrsZJGIgauPIENx0v10INIyFjmSNUD/gSqQ=="
     },
     "@babel/plugin-external-helpers": {
-      "version": "7.8.3",
```
Comment (saghul) [inline package-lock.json:1053]: I forgot to mention it earlier, please drop these changes. We usually update dependencies as dedicated PRs, unless a new fix / feature depends on the update.

File: react/features/app/middlewares.native.js [hunk 18]
```diff
@@ -13,6 +13,7 @@ import '../mobile/proximity/middleware';
 import '../mobile/wake-lock/middleware';
 import '../mobile/watchos/middleware';
 import '../share-room/middleware';
+import '../shared-video/middleware';
 import '../youtube-player/middleware';
```
Comment (saghul) [inline react/features/app/middlewares.native.js:17]: This one needs to go, right?

File: react/features/shared-video/components/web/SharedVideoButton.js [hunk 19]
```diff
@@ -0,0 +1,144 @@
+// @flow
+
+import type { Dispatch } from 'redux';
+
+import { isSharingStatus } from '../../';
```
Comment (saghul) [inline react/features/shared-video/components/web/SharedVideoButton.js:5]: can you make this import explicit please?

File: react/features/shared-video/index.js [hunk 20]
```diff
@@ -1,2 +1,5 @@
 export * from './actions';
 export * from './actionTypes';
+export * from './components';
```
Comment (saghul) [inline react/features/shared-video/index.js:3]: let's drop index.js while we're here, instead of adding more to it...

**Review by saghul — COMMENTED** [review]
Some minor comments, PTAL!

File: react/features/app/middlewares.native.js (hunk 18 above)
Comment (Calinteodor) [reply react/features/app/middlewares.native.js:17]: I thought about this, but wouldn't it be safer if we do this after we remove the whole youtube-player feature from the project structure?

Comment (saghul) [reply react/features/app/middlewares.native.js:17]: What do we have still there? Can't we move it to shared-video ?

File: package-lock.json (hunk 17 above)
Comment (Calinteodor) [reply package-lock.json]: This has been reverted.

File: react/features/app/middlewares.native.js (hunk 18 above)
Comment (Calinteodor) [reply react/features/app/middlewares.native.js:17]: We have some styles and a YoutubeLargeVideo component that is used in ParticipantView.native.js.

File: react/features/shared-video/components/web/SharedVideoButton.js (hunk 19 above)
Comment (Calinteodor) [reply react/features/shared-video/components/web/SharedVideoButton.js]: Done.

File: react/features/shared-video/index.js (hunk 20 above)
Comment (Calinteodor) [reply react/features/shared-video/index.js:3]: File contains only these 3 exports right now.

Comment (saghul) [reply react/features/shared-video/index.js:3]: kill the entire file and fix the imports please :-)

File: react/features/app/middlewares.native.js (hunk 18 above)
Comment (saghul) [reply react/features/app/middlewares.native.js]: Move that to the native components section of shared-video and fix the imports please

File: react/features/shared-video/index.js (hunk 20 above)
Comment (Calinteodor) [reply react/features/shared-video/index.js]: Killed

File: react/features/app/middlewares.native.js (hunk 18 above)
Comment (Calinteodor) [reply react/features/app/middlewares.native.js]: Done.

**Comment by saghul** [comment]
Jenkins please test this please.

**Comment by damencho** [comment]
Jenkins test this please.

File: react/features/shared-video/middleware.web.js [hunk 21]
```diff
@@ -0,0 +1,58 @@
...
+        if (conference && conference !== previousConference) {
+            conference.addCommandListener(SHARED_VIDEO,
+                ({ attributes }) => {
+
+                    const { dispatch, getState } = store;
+                    const { from } = attributes;
+                    const localParticipantId = getLocalParticipant(getState()).id;
+
+                    if (localParticipantId !== from) {
+                        dispatch(setDisableButton(true));
```
Comment (saghul) [inline react/features/shared-video/middleware.web.js:52]: When is this reset to false?

File: react/features/shared-video/middleware.web.js [hunk 22]
```diff
@@ -0,0 +1,58 @@
...
+    }
+
+    return next(action);
+});
+
+/**
+ * Set up state change listener to disable or enable the share video button in
+ * the toolbar menu.
+ */
+StateListenerRegistry.register(
```
Comment (saghul) [inline react/features/shared-video/middleware.web.js:40]: Nice!

File: react/features/shared-video/reducer.native.js [hunk 23]
```diff
@@ -6,13 +8,17 @@ import { SET_SHARED_VIDEO_STATUS } from './actionTypes';
...
  */
 ReducerRegistry.register('features/shared-video', (state = {}, action) => {
+    const { videoId, status, time, ownerId } = action;
+
     switch (action.type) {
     case SET_SHARED_VIDEO_STATUS:
         return {
             ...state,
-            status: action.status
+            videoId,
```
Comment (saghul) [inline react/features/shared-video/reducer.native.js:17]: This is the mobile reducer, why this change?

File: react/features/shared-video/reducer.web.js [hunk 24]
```diff
@@ -0,0 +1,29 @@
...
+import { SET_SHARED_VIDEO_STATUS, SET_DISABLE_BUTTON } from './actionTypes';
+
+/**
+ * Reduces the Redux actions of the feature features/shared-video.
+ */
+ReducerRegistry.register('features/shared-video', (state = {}, action) => {
+    const { status, disabled } = action;
+
+    switch (action.type) {
+    case SET_SHARED_VIDEO_STATUS:
```
Comment (saghul) [inline react/features/shared-video/reducer.web.js:14]: Is this emitted on web?

**Review by saghul — COMMENTED** [review]
Left some comments, PTAL.

File: react/features/shared-video/reducer.native.js (hunk 23 above)
Comment (Calinteodor) [reply react/features/shared-video/reducer.native.js:17]: I just used destructuring for the status that's it.

File: react/features/shared-video/reducer.web.js (hunk 24 above)
Comment (Calinteodor) [reply react/features/shared-video/reducer.web.js:14]: Yes.

File: react/features/shared-video/middleware.native.js [hunk 25]
```diff
@@ -70,7 +70,7 @@ StateListenerRegistry.register(
                     const localParticipantId = getLocalParticipant(getState()).id;
                     const status = attributes.state;
 
-                    if ([ 'playing', 'pause', 'start' ].includes(status)) {
+                    if (isSharingStatus(status)) {
```
Comment (Calinteodor) [inline react/features/shared-video/middleware.native.js:73]: Here I thought it would be better and cleaner if we use the helper function.

Comment (saghul) [reply react/features/shared-video/middleware.native.js:73]: 👍

**Review by saghul — DISMISSED** [review]
LGTM. Outstanding job @Calinteodor ! 🚀

**Review by saghul — APPROVED** [review]

File: react/features/shared-video/middleware.web.js (hunk 21 above)
Comment (Calinteodor) [reply react/features/shared-video/middleware.web.js]: Forgot to answer this. 
When the video is stopped by the owner, disabled is set to false and anyone in the conference can share a video.
